File: Backend/sms-backend/src/schemas/tenant/tenant.py
from typing import Optional, Dict, Any
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, Field, field_validator

from src.schemas.base.base import BaseSchema, TimestampSchema
from src.schemas.tenant.tenant_settings import TenantSettings


class TenantBase(BaseModel):
    """Base schema for Tenant model."""
    name: str
    code: str
    is_active: bool = True
    # created_at: Optional[datetime]
    # updated_at: Optional[datetime]
    domain: Optional[str] = None
    subdomain: Optional[str] = None
    logo: Optional[str] = None
    primary_color: Optional[str] = None  
    secondary_color: Optional[str] = None  
    
    @field_validator('code')
    def code_to_uppercase(cls, v):
        return v.upper() if v else v
    
    @field_validator('name')
    def validate_name_length(cls, v):
        if len(v) < 3:
            raise ValueError("tenant name must be at least 3 characters long")
        return v


class TenantCreate(TenantBase):
    """Schema for creating a new tenant."""
    pass


class AdminUserData(BaseModel):
    """Schema for admin user data when creating a tenant"""
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    email: str = Field(..., max_length=255)
    password: Optional[str] = Field(None, min_length=8)  # If None, password will be generated


class TenantCreateWithAdmin(TenantBase):
    """Schema for creating a tenant with admin user in a single atomic operation"""
    admin_user: AdminUserData


class TenantUpdate(BaseModel):
    """Schema for updating a tenant."""
    name: Optional[str] = None
    code: Optional[str] = None
    is_active: Optional[bool] = None
    # updated_at: Optional[datetime] = None
    domain: Optional[str] = None
    subdomain: Optional[str] = None
    logo: Optional[str] = None
    primary_color: Optional[str] = None  
    secondary_color: Optional[str] = None  
    
    @field_validator('code')
    def code_to_uppercase(cls, v):
        return v.upper() if v else v
    
    @field_validator('name')
    def validate_name_length(cls, v):
        if v and len(v) < 3:
            raise ValueError("tenant name must be at least 3 characters long")
        return v


class TenantCreateResponse(BaseModel):
    """Response schema for tenant creation with admin user"""
    tenant: 'Tenant'
    admin_user: dict  # Contains user info and generated password if applicable
    
    class Config:
        from_attributes = True


class Tenant(TenantBase, TimestampSchema):
    """Schema for Tenant model response."""
    id: UUID
    
    class Config:
        from_attributes = True
        populate_by_name = True


        

File: Frontend/sms-client/src/types/tenant.ts
export interface Tenant {
  id: string;
  name: string;
  code?: string;
  domain?: string | null;
  subdomain?: string | null;
  logo?: string | null;
  primaryColor?: string | null;
  secondaryColor?: string | null;
  isActive?: boolean;
  createdAt?: string;
  updatedAt?: string;
  plan_type?: string;
  plan_amount?: number;
  subscription_status?: string;
  features?: {
    enable_finance: boolean;
    enable_transportation: boolean;
    enable_cafeteria: boolean;
    enable_health: boolean;
    enable_parent_portal: boolean;
    enable_sms_notifications: boolean;
    [key: string]: boolean;
  };
}

export interface TenantCreate {
  name: string;
  code: string;
  domain?: string;
  subdomain?: string;
  logo?: string;
  primaryColor?: string;
  secondaryColor?: string;
  isActive?: boolean;
  created_at?: string;
  updated_at?: string;
  plan_type?: string;
  plan_amount?: number;
  subscription_status?: string;
}

export interface TenantUpdate {
  name?: string;
  code?: string;
  domain?: string;
  subdomain?: string;
  logo?: string;
  primaryColor?: string;
  secondaryColor?: string;
  isActive?: boolean;
  plan_type?: string;
  plan_amount?: number;
  subscription_status?: string;
}

export interface AdminUserData {
  first_name: string;
  last_name: string;
  email: string;
  password?: string; // If not provided, password will be generated
}

export interface TenantCreateWithAdmin {
  name: string;
  code: string;
  domain?: string;
  subdomain?: string;
  logo?: string;
  primaryColor?: string;
  secondaryColor?: string;
  isActive?: boolean;
  admin_user: AdminUserData;
}

export interface TenantCreateResponse {
  tenant: Tenant;
  admin_user: {
    id: string;
    email: string;
    first_name: string;
    last_name: string;
    is_active: boolean;
    created_at: string;
    generated_password?: string;
  };
}